crates/pallet-messenger/src/verification.rs

```diff
@@ -0,0 +1,48 @@
+use codec::{Decode, Encode};
+use frame_support::PalletError;
+use hash_db::Hasher;
+use scale_info::TypeInfo;
+use sp_core::storage::StorageKey;
+use sp_trie::{read_trie_value, LayoutV1, StorageProof};
+use std::marker::PhantomData;
+
+/// Verification error.
+#[derive(Debug, PartialEq, Eq, Encode, Decode, PalletError, TypeInfo)]
+pub enum VerificationError {
+    /// Emits when the expected state root doesn't exist
+    InvalidStateRoot,
+    /// Emits when the given storage proof is invalid.
+    InvalidProof,
+    /// Value doesn't exist in the Db for the given key.
+    MissingValue,
+    /// Failed to decode value.
+    FailedToDecode,
+}
+
+pub(crate) struct StorageProofVerifier<H: Hasher>(PhantomData<H>);
+
+impl<H: Hasher> StorageProofVerifier<H> {
+    pub(crate) fn verify_and_get_value<V: Decode>(
+        proof: Proof<H::Out>,
+        key: StorageKey,
+    ) -> Result<V, VerificationError> {
+        let db = proof.message_proof.into_memory_db::<H>();
+        let val = read_trie_value::<LayoutV1<H>, _>(&db, &proof.state_root, key.as_ref())
+            .map_err(|_| VerificationError::InvalidProof)?
+            .ok_or(VerificationError::MissingValue)?;
+
+        let decoded = V::decode(&mut &val[..]).map_err(|_| VerificationError::FailedToDecode)?;
+
+        Ok(decoded)
+    }
+}
+
+/// Proof combines the storage proofs to validate messages.
+#[derive(Debug, Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+pub struct Proof<StateRoot> {
+    pub state_root: StateRoot,
+    /// Storage proof that src_domain state_root is registered on System domain
+    // TODO(ved): add system domain proof when store is available
+    /// Storage proof that message is processed on src_domain.
+    pub message_proof: StorageProof,
```

Comment:
Could you explain a bit more how `message_proof` will be actually generated? Now I'm a little confused after looking at the usage of `SystemDomainTracker`.

Reply:
Take a look at how i'm generating it here - https://github.com/subspace/subspace/pull/844/files#diff-04683350ee351b2896694b92a1120a12fc9427a16e39a71385ba12d068eb7ce8R120

Construct the trie and decode the value using the constructed key from verifier point of view. For message to be verified, on dst_domain we reconstruct the key from point of view of src_domain i.e message for dst_domain channelId and nonce is present in Outbox of src_domain and the state_root of src_domain is decoded from the missing(TODO) storage proof that can verify that state_root of the src_domain is part of the System domain state.

So the state_root field in the Proof is the state root of the Syatem domain. Verifier will fetch the last n K-deep confirmed state roots of System domain. This will allow relayers to not generate Proof everytime state root is updated

Reply:
Thanks for the explanation!